File: examtopics_helper/question_set.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Optional


@dataclass(frozen=True)
class Choice:
    id: str
    text: str


@dataclass(frozen=True)
class Question:
    id: str
    text: str
    choices: list[Choice]
    answer_choice_ids: Optional[list[str]] = None
    # Explicitly indicates whether the UI should allow multiple selections.
    # If omitted, the app may infer from answer_choice_ids length / "(Choose two.)" text.
    is_multi_select: Optional[bool] = None
    explanation: Optional[str] = None
    tags: Optional[list[str]] = None


@dataclass(frozen=True)
class QuestionSet:
    set_id: str
    title: str
    questions: list[Question]


class QuestionSetFormatError(ValueError):
    pass


def _require(d: dict[str, Any], k: str) -> Any:
    if k not in d:
        raise QuestionSetFormatError(f"missing required field: {k}")
    return d[k]


def _as_str(x: Any, *, field: str) -> str:
    if not isinstance(x, str) or not x.strip():
        raise QuestionSetFormatError(f"{field} must be a non-empty string")
    return x


def _as_optional_str(x: Any, *, field: str) -> Optional[str]:
    if x is None:
        return None
    if not isinstance(x, str):
        raise QuestionSetFormatError(f"{field} must be a string or null")
    s = x.strip()
    return s if s else None


def _as_optional_str_list(x: Any, *, field: str) -> Optional[list[str]]:
    if x is None:
        return None
    if not isinstance(x, list) or not all(isinstance(i, str) for i in x):
        raise QuestionSetFormatError(f"{field} must be a list of strings or null")
    out = [i.strip() for i in x if i.strip()]
    return out or None


def _as_optional_bool(x: Any, *, field: str) -> Optional[bool]:
    if x is None:
        return None
    if not isinstance(x, bool):
        raise QuestionSetFormatError(f"{field} must be a boolean or null")
    return x


def parse_question_set(obj: Any) -> QuestionSet:
    if not isinstance(obj, dict):
        raise QuestionSetFormatError("root must be an object")

    set_id = _as_str(_require(obj, "set_id"), field="set_id")
    title = _as_str(_require(obj, "title"), field="title")
    questions_raw = _require(obj, "questions")
    if not isinstance(questions_raw, list):
        raise QuestionSetFormatError("questions must be a list")

    seen_qids: set[str] = set()
    questions: list[Question] = []
    for i, q in enumerate(questions_raw, start=1):
        if not isinstance(q, dict):
            raise QuestionSetFormatError(f"questions[{i}] must be an object")

        qid = _as_str(_require(q, "id"), field=f"questions[{i}].id")
        if qid in seen_qids:
            raise QuestionSetFormatError(f"duplicate question id: {qid}")
        seen_qids.add(qid)

        qtext = _as_str(_require(q, "text"), field=f"questions[{i}].text")
        choices_raw = _require(q, "choices")
        if not isinstance(choices_raw, list) or not choices_raw:
            raise QuestionSetFormatError(f"questions[{i}].choices must be a non-empty list")

        seen_cids: set[str] = set()
        choices: list[Choice] = []
        for j, c in enumerate(choices_raw, start=1):
            if not isinstance(c, dict):
                raise QuestionSetFormatError(f"questions[{i}].choices[{j}] must be an object")
            cid = _as_str(_require(c, "id"), field=f"questions[{i}].choices[{j}].id")
            if cid in seen_cids:
                raise QuestionSetFormatError(f"duplicate choice id in question {qid}: {cid}")
            seen_cids.add(cid)
            ctext = _as_str(_require(c, "text"), field=f"questions[{i}].choices[{j}].text")
            choices.append(Choice(id=cid, text=ctext))

        answer_choice_ids = _as_optional_str_list(q.get("answer_choice_ids"), field=f"questions[{i}].answer_choice_ids")
        if answer_choice_ids:
            unknown = [cid for cid in answer_choice_ids if cid not in seen_cids]
            if unknown:
                raise QuestionSetFormatError(
                    f"questions[{i}].answer_choice_ids contains unknown choice ids: {unknown}"
                )

        is_multi_select = _as_optional_bool(q.get("is_multi_select"), field=f"questions[{i}].is_multi_select")

        explanation = _as_optional_str(q.get("explanation"), field=f"questions[{i}].explanation")
        tags = _as_optional_str_list(q.get("tags"), field=f"questions[{i}].tags")

        questions.append(
            Question(
                id=qid,
                text=qtext,
                choices=choices,
                answer_choice_ids=answer_choice_ids,
                is_multi_select=is_multi_select,
                explanation=explanation,
                tags=tags,
            )
        )

    return QuestionSet(set_id=set_id, title=title, questions=questions)


def load_question_set_json_bytes(data: bytes) -> QuestionSet:
    try:
        obj = json.loads(data.decode("utf-8"))
    except Exception as e:  # pragma: no cover
        raise QuestionSetFormatError(f"failed to parse JSON: {e}") from e
    return parse_question_set(obj)


def load_question_set_file(path: str) -> QuestionSet:
    with open(path, "rb") as f:
        return load_question_set_json_bytes(f.read())

File: web/src/lib/questionSet.ts
export type Choice = {
  id: string;
  text: string;
};

export type Question = {
  id: string;
  text: string;
  choices: Choice[];
  answer_choice_ids?: string[] | null;
  is_multi_select?: boolean | null;
  explanation?: string | null;
  tags?: string[] | null;
};

export type QuestionSet = {
  set_id: string;
  title: string;
  questions: Question[];
};

export class QuestionSetFormatError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "QuestionSetFormatError";
  }
}

function requireField(obj: Record<string, unknown>, key: string): unknown {
  if (!(key in obj)) throw new QuestionSetFormatError(`missing required field: ${key}`);
  return obj[key];
}

function asNonEmptyString(x: unknown, field: string): string {
  if (typeof x !== "string" || x.trim().length === 0) {
    throw new QuestionSetFormatError(`${field} must be a non-empty string`);
  }
  return x;
}

function asOptionalString(x: unknown, field: string): string | null | undefined {
  if (x === undefined) return undefined;
  if (x === null) return null;
  if (typeof x !== "string") throw new QuestionSetFormatError(`${field} must be a string or null`);
  const s = x.trim();
  return s.length ? s : null;
}

function asOptionalStringList(x: unknown, field: string): string[] | null | undefined {
  if (x === undefined) return undefined;
  if (x === null) return null;
  if (!Array.isArray(x) || !x.every((i) => typeof i === "string")) {
    throw new QuestionSetFormatError(`${field} must be a list of strings or null`);
  }
  const out = x.map((i) => i.trim()).filter(Boolean);
  return out.length ? out : null;
}

function asOptionalBoolean(x: unknown, field: string): boolean | null | undefined {
  if (x === undefined) return undefined;
  if (x === null) return null;
  if (typeof x !== "boolean") throw new QuestionSetFormatError(`${field} must be a boolean or null`);
  return x;
}

export function parseQuestionSet(obj: unknown): QuestionSet {
  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
    throw new QuestionSetFormatError("root must be an object");
  }
  const root = obj as Record<string, unknown>;

  const set_id = asNonEmptyString(requireField(root, "set_id"), "set_id");
  const title = asNonEmptyString((root.title as unknown) ?? set_id, "title");

  const questionsRaw = requireField(root, "questions");
  if (!Array.isArray(questionsRaw)) throw new QuestionSetFormatError("questions must be a list");

  const seenQids = new Set<string>();
  const questions: Question[] = questionsRaw.map((q, idx0) => {
    const i = idx0 + 1;
    if (typeof q !== "object" || q === null || Array.isArray(q)) {
      throw new QuestionSetFormatError(`questions[${i}] must be an object`);
    }
    const qObj = q as Record<string, unknown>;

    const qid = asNonEmptyString(requireField(qObj, "id"), `questions[${i}].id`);
    if (seenQids.has(qid)) throw new QuestionSetFormatError(`duplicate question id: ${qid}`);
    seenQids.add(qid);

    const text = asNonEmptyString(requireField(qObj, "text"), `questions[${i}].text`);

    const choicesRaw = requireField(qObj, "choices");
    if (!Array.isArray(choicesRaw) || choicesRaw.length === 0) {
      throw new QuestionSetFormatError(`questions[${i}].choices must be a non-empty list`);
    }

    const seenCids = new Set<string>();
    const choices: Choice[] = choicesRaw.map((c, idx1) => {
      const j = idx1 + 1;
      if (typeof c !== "object" || c === null || Array.isArray(c)) {
        throw new QuestionSetFormatError(`questions[${i}].choices[${j}] must be an object`);
      }
      const cObj = c as Record<string, unknown>;
      const cid = asNonEmptyString(
        requireField(cObj, "id"),
        `questions[${i}].choices[${j}].id`,
      );
      if (seenCids.has(cid)) throw new QuestionSetFormatError(`duplicate choice id in question ${qid}: ${cid}`);
      seenCids.add(cid);
      const ctext = asNonEmptyString(
        requireField(cObj, "text"),
        `questions[${i}].choices[${j}].text`,
      );
      return { id: cid, text: ctext };
    });

    const answer_choice_ids = asOptionalStringList(
      qObj.answer_choice_ids,
      `questions[${i}].answer_choice_ids`,
    );
    if (answer_choice_ids && answer_choice_ids.length) {
      const unknown = answer_choice_ids.filter((cid) => !seenCids.has(cid));
      if (unknown.length) {
        throw new QuestionSetFormatError(
          `questions[${i}].answer_choice_ids contains unknown choice ids: ${unknown.join(", ")}`,
        );
      }
    }

    const is_multi_select = asOptionalBoolean(qObj.is_multi_select, `questions[${i}].is_multi_select`);
    const explanation = asOptionalString(qObj.explanation, `questions[${i}].explanation`);
    const tags = asOptionalStringList(qObj.tags, `questions[${i}].tags`);

    return {
      id: qid,
      text,
      choices,
      answer_choice_ids: answer_choice_ids ?? undefined,
      is_multi_select: is_multi_select ?? undefined,
      explanation: explanation ?? undefined,
      tags: tags ?? undefined,
    };
  });

  return { set_id, title, questions };
}

export function loadQuestionSetFromJsonText(jsonText: string): QuestionSet {
  let obj: unknown;
  try {
    obj = JSON.parse(jsonText);
  } catch (e) {
    throw new QuestionSetFormatError(`failed to parse JSON: ${String(e)}`);
  }
  return parseQuestionSet(obj);
}
